pkg/applicationserver/io/nats/nats.go

```diff
@@ -0,0 +1,168 @@
+	"go.thethings.network/lorawan-stack/pkg/unique"
+)
+
+const (
+	wildcard         = "*"
+	tailWildcard     = ">"
+	subjectSeparator = "."
+)
+
+type srv struct {
+	ctx    context.Context
+	server io.Server
+	format iopubsub.Format
+	conn   *nats.Conn
+	subs   []*nats.Subscription
+}
+
+// Start starts the nats frontend.
+func Start(ctx context.Context, server io.Server, format iopubsub.Format, natsServer string) (sub *io.Subscription, err error) {
+	ctx = log.NewContextWithField(ctx, "namespace", "applicationserver/io/nats")
+	conn, err := nats.Connect(natsServer)
+	if err != nil {
+		return nil, err
+	}
+	s := &srv{
+		ctx:    ctx,
+		server: server,
+		format: format,
+		conn:   conn,
+	}
+	go func() {
+		<-ctx.Done()
+		for _, sub := range s.subs {
+			sub.Unsubscribe()
+		}
+		s.conn.Close()
+	}()
+
+	sub = io.NewSubscription(s.ctx, "nats", nil)
+	// Publish upstream
+	go func() {
+		logger := log.FromContext(s.ctx)
+		for {
+			select {
+			case <-sub.Context().Done():
+				logger.WithError(sub.Context().Err()).Debug("Done sending upstream messages")
+				return
+			case up := <-sub.Up():
+				appID := up.ApplicationIdentifiers
+				var subjectParts []string
+				switch up.Up.(type) {
+				case *ttnpb.ApplicationUp_UplinkMessage:
+					subjectParts = s.format.UplinkTopic(unique.ID(sub.Context(), appID), up.DeviceID)
+				case *ttnpb.ApplicationUp_JoinAccept:
+					subjectParts = s.format.JoinAcceptTopic(unique.ID(sub.Context(), appID), up.DeviceID)
+				case *ttnpb.ApplicationUp_DownlinkAck:
+					subjectParts = s.format.DownlinkAckTopic(unique.ID(sub.Context(), appID), up.DeviceID)
+				case *ttnpb.ApplicationUp_DownlinkNack:
+					subjectParts = s.format.DownlinkNackTopic(unique.ID(sub.Context(), appID), up.DeviceID)
+				case *ttnpb.ApplicationUp_DownlinkSent:
+					subjectParts = s.format.DownlinkSentTopic(unique.ID(sub.Context(), appID), up.DeviceID)
+				case *ttnpb.ApplicationUp_DownlinkFailed:
+					subjectParts = s.format.DownlinkFailedTopic(unique.ID(sub.Context(), appID), up.DeviceID)
+				case *ttnpb.ApplicationUp_DownlinkQueued:
+					subjectParts = s.format.DownlinkQueuedTopic(unique.ID(sub.Context(), appID), up.DeviceID)
+				case *ttnpb.ApplicationUp_LocationSolved:
+					subjectParts = s.format.LocationSolvedTopic(unique.ID(sub.Context(), appID), up.DeviceID)
+				}
+				if subjectParts == nil {
+					continue
+				}
+				buf, err := s.format.FromUp(up)
+				if err != nil {
+					log.WithError(err).Warn("Failed to marshal upstream message")
+					continue
+				}
+				subject := createSubject(subjectParts)
+				logger.Infof("Publish upstream message to subject %v", subject)
+				s.conn.Publish(subject, buf)
+			}
+		}
+	}()
+
+	// Subscribe downstream
+	for _, subjectParts := range [][]string{
+		s.format.DownlinkPushTopic(wildcard, wildcard),
+		s.format.DownlinkReplaceTopic(wildcard, wildcard),
+	} {
+		nsub, err := conn.Subscribe(createSubject(subjectParts), func(m *nats.Msg) {
```

Comment:
Please use `QueueSubscribe` so that we won't run into surprises when running multiple AS instances.